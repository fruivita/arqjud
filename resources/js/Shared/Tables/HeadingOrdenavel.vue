<!--
    Heading (th) com função de ordenação.

    Notar que:
    - Notar que o componente depende do status do processamento geral da
    aplicação para definir seu próprio estado, por exemplo, caso a página
    esteja sendo processada, os botões são desabilitados;
    - Se o usuário clicar no botão de ordenação, um evento é emitido com o
    respectivo valor de ordenação (asc, desc ou undefined). Apenas para efeito
    estético, o ícone é alterado.

    @link https://vuejs.org/guide/introduction.html
    @link https://tailwindcss.com/docs
    @link https://inertiajs.com/
 -->

<script setup>
import Icone from '@/Shared/Icones/Icone.vue';
import { useStatusRequisicaoStore } from '@/Stores/StatusRequisicaoStore';
import { computed } from 'vue';

const props = defineProps({
    texto: { type: String, required: true },
    ordenacao: { type: String },
});

const emit = defineEmits(['ordenar']);

const status = useStatusRequisicaoStore();

const icone = computed(function () {
    if (props.ordenacao === 'asc') {
        return 'arrow-up-short';
    }
    if (props.ordenacao === 'desc') {
        return 'arrow-down-short';
    }
    return 'dash';
});

const setOrdenacao = () => {
    let ordenacao;

    if (props.ordenacao === 'asc') {
        ordenacao = 'desc';
    } else if (props.ordenacao === 'desc') {
        ordenacao = undefined;
    } else {
        ordenacao = 'asc';
    }

    emit('ordenar', ordenacao);
};
</script>

<template>
    <th class="p-3">
        <button
            :disabled="status.processando === true"
            @click="setOrdenacao"
            class="transform-gpu whitespace-nowrap rounded px-3 outline-none transition duration-300 hover:bg-primaria-300 focus:bg-primaria-300 disabled:cursor-not-allowed hover:dark:bg-secundaria-500 focus:dark:bg-secundaria-500"
            dusk="ordenacao"
        >
            <span class="font-bold">{{ texto }}</span>

            <Transition
                enter-from-class="rotate-180"
                enter-to-class="rotate-0"
                enter-active-class="transition duration-300 transform-gpu"
                leave-active-class="transition duration-300 transform-gpu"
                leave-from-class="rotate-0"
                leave-to-class="rotate-180"
            >
                <Icone :nome="icone" class="inline h-6 w-6" />
            </Transition>
        </button>
    </th>
</template>
