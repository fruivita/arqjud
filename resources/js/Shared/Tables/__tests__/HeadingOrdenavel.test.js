/**
 * Testes para o componente HeadingOrdenavel.
 *
 * @link https://vuejs.org/guide/scaling-up/testing.html
 * @link https://test-utils.vuejs.org/guide/
 * @link https://vitest.dev/
 */

import HeadingOrdenavel from '@/Shared/Tables/HeadingOrdenavel.vue';
import { createTestingPinia } from '@pinia/testing';
import { mount } from '@vue/test-utils';
import { beforeEach, describe, expect, test } from 'vitest';

let mountFunction;
let pinia;

beforeEach(() => {
    pinia = (options = {}) =>
        createTestingPinia({
            initialState: {
                StatusRequisicaoStore: { ...options },
            },
        });

    mountFunction = (options = {}) => {
        return mount(HeadingOrdenavel, { ...options });
    };
});

// Caminho feliz
describe('HeadingOrdenavel', () => {
    test('propriedades do componente estão definidas', () => {
        expect(HeadingOrdenavel.props).toMatchObject({
            texto: { type: String, required: true },
            ordenacao: { type: String },
        });
    });

    const ordenacoes = [
        ['', 'asc'],
        ['asc', 'desc'],
        ['desc', undefined],
    ];
    test.each(ordenacoes)(
        'renderiza o componente com o ícone respectivo respeitando o snapshot',
        (ordenacao) => {
            expect(
                mountFunction({
                    props: { texto: 'foo', ordenacao: ordenacao },
                    global: {
                        plugins: [pinia({ processando: false })],
                    },
                }).html()
            ).toMatchSnapshot();
        }
    );

    test.each(ordenacoes)(
        'renderiza o componente de maneira diversa enquanto a página se carrega respeitando o snapshot',
        (entrada) => {
            expect(
                mountFunction({
                    props: { texto: 'foo', ordenacao: entrada },
                    global: {
                        plugins: [pinia({ processando: true })],
                    },
                }).html()
            ).toMatchSnapshot();
        }
    );

    test.each(ordenacoes)(
        'dispara o evento ordenar com o novo valor da ordenação ao clicar no componente',
        (entrada, saida) => {
            const wrapper = mountFunction({
                props: { texto: 'foo', ordenacao: entrada },
                global: {
                    plugins: [pinia({ processando: false })],
                },
            });

            wrapper.vm.setOrdenacao();

            expect(wrapper.emitted()['ordenar'][0]).toEqual([saida]);
        }
    );
});
