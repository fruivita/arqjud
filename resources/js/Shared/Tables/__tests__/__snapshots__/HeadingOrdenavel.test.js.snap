// Vitest Snapshot v1

exports[`HeadingOrdenavel > renderiza o componente com o ícone respectivo respeitando o snapshot 1`] = `
"<th class=\\"p-3\\"><button class=\\"transform-gpu whitespace-nowrap rounded px-3 outline-none transition duration-300 hover:bg-primaria-300 focus:bg-primaria-300 disabled:cursor-not-allowed hover:dark:bg-secundaria-500 focus:dark:bg-secundaria-500\\" dusk=\\"ordenacao\\"><span class=\\"font-bold\\">foo</span>
    <transition-stub enterfromclass=\\"rotate-180\\" entertoclass=\\"rotate-0\\" enteractiveclass=\\"transition duration-300 transform-gpu\\" leaveactiveclass=\\"transition duration-300 transform-gpu\\" leavefromclass=\\"rotate-0\\" leavetoclass=\\"rotate-180\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\"><svg class=\\"inline h-6 w-6\\" fill=\\"currentColor\\" viewBox=\\"0 0 16 16\\" xmlns=\\"http://www.w3.org/2000/svg\\">
        <path d=\\"M4 8a.5.5 0 0 1 .5-.5h7a.5.5 0 0 1 0 1h-7A.5.5 0 0 1 4 8z\\"></path>
      </svg></transition-stub>
  </button></th>"
`;

exports[`HeadingOrdenavel > renderiza o componente com o ícone respectivo respeitando o snapshot 2`] = `
"<th class=\\"p-3\\"><button class=\\"transform-gpu whitespace-nowrap rounded px-3 outline-none transition duration-300 hover:bg-primaria-300 focus:bg-primaria-300 disabled:cursor-not-allowed hover:dark:bg-secundaria-500 focus:dark:bg-secundaria-500\\" dusk=\\"ordenacao\\"><span class=\\"font-bold\\">foo</span>
    <transition-stub enterfromclass=\\"rotate-180\\" entertoclass=\\"rotate-0\\" enteractiveclass=\\"transition duration-300 transform-gpu\\" leaveactiveclass=\\"transition duration-300 transform-gpu\\" leavefromclass=\\"rotate-0\\" leavetoclass=\\"rotate-180\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\"><svg class=\\"inline h-6 w-6\\" fill=\\"currentColor\\" viewBox=\\"0 0 16 16\\" xmlns=\\"http://www.w3.org/2000/svg\\">
        <path fill-rule=\\"evenodd\\" d=\\"M8 12a.5.5 0 0 0 .5-.5V5.707l2.146 2.147a.5.5 0 0 0 .708-.708l-3-3a.5.5 0 0 0-.708 0l-3 3a.5.5 0 1 0 .708.708L7.5 5.707V11.5a.5.5 0 0 0 .5.5z\\"></path>
      </svg></transition-stub>
  </button></th>"
`;

exports[`HeadingOrdenavel > renderiza o componente com o ícone respectivo respeitando o snapshot 3`] = `
"<th class=\\"p-3\\"><button class=\\"transform-gpu whitespace-nowrap rounded px-3 outline-none transition duration-300 hover:bg-primaria-300 focus:bg-primaria-300 disabled:cursor-not-allowed hover:dark:bg-secundaria-500 focus:dark:bg-secundaria-500\\" dusk=\\"ordenacao\\"><span class=\\"font-bold\\">foo</span>
    <transition-stub enterfromclass=\\"rotate-180\\" entertoclass=\\"rotate-0\\" enteractiveclass=\\"transition duration-300 transform-gpu\\" leaveactiveclass=\\"transition duration-300 transform-gpu\\" leavefromclass=\\"rotate-0\\" leavetoclass=\\"rotate-180\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\"><svg class=\\"inline h-6 w-6\\" fill=\\"currentColor\\" viewBox=\\"0 0 16 16\\" xmlns=\\"http://www.w3.org/2000/svg\\">
        <path fill-rule=\\"evenodd\\" d=\\"M8 4a.5.5 0 0 1 .5.5v5.793l2.146-2.147a.5.5 0 0 1 .708.708l-3 3a.5.5 0 0 1-.708 0l-3-3a.5.5 0 1 1 .708-.708L7.5 10.293V4.5A.5.5 0 0 1 8 4z\\"></path>
      </svg></transition-stub>
  </button></th>"
`;

exports[`HeadingOrdenavel > rendiza o componente de maneira diversa enquanto a página se carrega respeitando o snapshot 1`] = `
"<th class=\\"p-3\\"><button disabled=\\"\\" class=\\"transform-gpu whitespace-nowrap rounded px-3 outline-none transition duration-300 hover:bg-primaria-300 focus:bg-primaria-300 disabled:cursor-not-allowed hover:dark:bg-secundaria-500 focus:dark:bg-secundaria-500\\" dusk=\\"ordenacao\\"><span class=\\"font-bold\\">foo</span>
    <transition-stub enterfromclass=\\"rotate-180\\" entertoclass=\\"rotate-0\\" enteractiveclass=\\"transition duration-300 transform-gpu\\" leaveactiveclass=\\"transition duration-300 transform-gpu\\" leavefromclass=\\"rotate-0\\" leavetoclass=\\"rotate-180\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\"><svg class=\\"inline h-6 w-6\\" fill=\\"currentColor\\" viewBox=\\"0 0 16 16\\" xmlns=\\"http://www.w3.org/2000/svg\\">
        <path d=\\"M4 8a.5.5 0 0 1 .5-.5h7a.5.5 0 0 1 0 1h-7A.5.5 0 0 1 4 8z\\"></path>
      </svg></transition-stub>
  </button></th>"
`;

exports[`HeadingOrdenavel > rendiza o componente de maneira diversa enquanto a página se carrega respeitando o snapshot 2`] = `
"<th class=\\"p-3\\"><button disabled=\\"\\" class=\\"transform-gpu whitespace-nowrap rounded px-3 outline-none transition duration-300 hover:bg-primaria-300 focus:bg-primaria-300 disabled:cursor-not-allowed hover:dark:bg-secundaria-500 focus:dark:bg-secundaria-500\\" dusk=\\"ordenacao\\"><span class=\\"font-bold\\">foo</span>
    <transition-stub enterfromclass=\\"rotate-180\\" entertoclass=\\"rotate-0\\" enteractiveclass=\\"transition duration-300 transform-gpu\\" leaveactiveclass=\\"transition duration-300 transform-gpu\\" leavefromclass=\\"rotate-0\\" leavetoclass=\\"rotate-180\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\"><svg class=\\"inline h-6 w-6\\" fill=\\"currentColor\\" viewBox=\\"0 0 16 16\\" xmlns=\\"http://www.w3.org/2000/svg\\">
        <path fill-rule=\\"evenodd\\" d=\\"M8 12a.5.5 0 0 0 .5-.5V5.707l2.146 2.147a.5.5 0 0 0 .708-.708l-3-3a.5.5 0 0 0-.708 0l-3 3a.5.5 0 1 0 .708.708L7.5 5.707V11.5a.5.5 0 0 0 .5.5z\\"></path>
      </svg></transition-stub>
  </button></th>"
`;

exports[`HeadingOrdenavel > rendiza o componente de maneira diversa enquanto a página se carrega respeitando o snapshot 3`] = `
"<th class=\\"p-3\\"><button disabled=\\"\\" class=\\"transform-gpu whitespace-nowrap rounded px-3 outline-none transition duration-300 hover:bg-primaria-300 focus:bg-primaria-300 disabled:cursor-not-allowed hover:dark:bg-secundaria-500 focus:dark:bg-secundaria-500\\" dusk=\\"ordenacao\\"><span class=\\"font-bold\\">foo</span>
    <transition-stub enterfromclass=\\"rotate-180\\" entertoclass=\\"rotate-0\\" enteractiveclass=\\"transition duration-300 transform-gpu\\" leaveactiveclass=\\"transition duration-300 transform-gpu\\" leavefromclass=\\"rotate-0\\" leavetoclass=\\"rotate-180\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\"><svg class=\\"inline h-6 w-6\\" fill=\\"currentColor\\" viewBox=\\"0 0 16 16\\" xmlns=\\"http://www.w3.org/2000/svg\\">
        <path fill-rule=\\"evenodd\\" d=\\"M8 4a.5.5 0 0 1 .5.5v5.793l2.146-2.147a.5.5 0 0 1 .708.708l-3 3a.5.5 0 0 1-.708 0l-3-3a.5.5 0 1 1 .708-.708L7.5 10.293V4.5A.5.5 0 0 1 8 4z\\"></path>
      </svg></transition-stub>
  </button></th>"
`;
